### ISSUE TITLE
Workbench: filter Without Decision works incorrectly

### ISSUE THREAD

**Issue by yroskov** [CatalogueOfLife/backend#729]
Frontend version: 0e742b8 May 19, 2020 6:24 AMBackend version: a51ee40 May 18, 2020 12:45 PM

Checklist of the Collembola of the World - Workbench (https://data.catalogue.life/catalogue/3/dataset/2130/workbench?facet=rank&facet=issue&facet=status&facet=nomstatus&facet=type&facet=field&issue=name%20match%20none&limit=50&offset=0&reverse=false)

I have tried to filter only issues without decision. I have got 0 records in a report:
![image](https://user-images.githubusercontent.com/28574252/82474272-6cd3a800-9a90-11ea-8788-81b6fe57bf35.png)

However, there are many records without decision in a report if I chosen option All:

![image](https://user-images.githubusercontent.com/28574252/82474767-27fc4100-9a91-11ea-91ea-f79bdd27b7f0.png)

**Comment by ayco-at-naturalis** [comment]
@thomasstjerne (and @mdoering)  I wonder what the request looks like that your fire when option "Without decisions" is enabled.

Actually as it stands there is no way to do this currently. The CATALOGUE_KEY search parameter, which corresponds to the decision's dataset key is currently not actually used for searching. It is only used for post-processing the query result - removing decisions from name usages that do not match the value of CATALOGUE_KEY.

In other words &catalogueKey=12345 does not give you name usages with a catalogue key of 12345. It only removes the decision with catalogue key 12345 from **all** name usages that came back from the search.

I will now introduce an exception for the value _IS_NULL and _IS_NOT_NULL. So &catalogKey=_NULL **will** add an extra query constrant to the Elasticsearch query itself.

**Comment by mdoering** [comment]
@ayco-at-naturalis please check your browser to see the API calls the frontend does. We do not use any proxy API, so you can see the direct calls nicely.

When I activate the without decisions switch the UI does the following request (ignoring all facet params):

`https://api.catalogue.life/dataset/2130/nameusage/search?catalogueKey=3&decisionMode=_NULL&issue=name%20match%20none&limit=50&offset=0&reverse=false`

The patch parameter `dataset/2130` gets translated into the `searchRequest.datasetKey=2130`.
This looks like correct behavior according to the current search parameter java docs to me:

```
  DATASET_KEY(Integer.class),

  /**
   * This takes the datasetKey of the managed catalogue to filter decisions by, not usages. It will prune the list of decisions to just the
   * ones matching the datasetKey. I.e. the list only contains a single decision at max if one catalogue is given so the UI can quickly
   * determine if a decision exists at all for a given usage and does not have to ignore decisions from other catalogues.
   */
  CATALOGUE_KEY(Integer.class),

  /**
   * Allows to filter usages based on the existence of a decision with the matching MODE. NOT_NULL or NULL can be used here to filter usages
   * to just the ones with or without any decision (from the requested DECISION_DATASET_KEY if given)
   */
  DECISION_MODE(EditorialDecision.Mode.class),
```

The backend does not seem to recognoze DECISION_MODE=NULL properly, is that right?
Is there anything wrong with the design we took? If not, please implement it that way instead of the changes you proposed above.

**Comment by ayco-at-naturalis** [comment]
@mdoering Ah, now I see the issue: **catalogueKey=3&decisionMode=_NULL**

In other words: the __is__ a decision (the name usage is "with decision"); only its decision mode needs to be null.

Therefore it seems like the solution I proposed is correct: we need to add to option to ask for

catalogueKey=_NULL

**Comment by ayco-at-naturalis** [comment]
Addendum: the only EXTRA options I proposed is: we might then as well ask for:

catalogueKey=_NULL (give all name usages WITHOUT a decision, as above)
catalogueKey=_NOT_NULL (give all name usages WITH a decision, whatever the mode
catalogueKey=3 (no decisionMode param) give all decision for catalog 3, whatever the mode

I don't know if these options make sense. This is how I have implemented it now.

Let me know if I should scrap any of these options

**Comment by mdoering** [comment]
Why not stick to the original design? Otherwise we also need to change the UI again.

 - catalogueKey does only filter decisions, never filtering usages
 - decisionMode filters usages based on the existence of a decision and its mode. I.e:
   - mode=NULL: no mode=no decision. Show only usages without a decision
   - mode=NOT_NULL: some mode=some decision. Show only usages with a decision
   - mode=XYZ: Show only usages with a decision of this specific type

The problem I guess is that catalogueKey has to be taken into account for the mode too.
If mode=NULL or NOT_NULL it only makes sense in the context of a given catalogue.

I cannot see how your solution making use of catalogueKey should work. The query asking for usages without any decision is always in the context of a catalogueKey. It should not matter if a different catalogue has a decision. So we cannot use the catalogueKey to pass the NULL/NOT_NULL value

**Comment by mdoering** [comment]
In the original design `/dataset/2130/nameusage/search?catalogueKey=3&decisionMode=NULL` would select all usages from dataset 2130 that has no decision bound to catalogue 3. It can very well have other decisions though!

**Comment by ayco-at-naturalis** [comment]
OK, I understand. My original impl was pretty off-piste then, but I fixed. Will just add some more tests

Just to be sure: since **any** decision always has **some** mode (mode is never null), it is save to use it as a means of establishing whether a name usage has any decision at all.

**Comment by mdoering** [comment]
yes, thats totally correct!

**Comment by mdoering** [comment]
I *think* we still need to adjust the translator code.
@ayco-at-naturalis I have updated the code. Could you review if it makes sense please?
https://github.com/CatalogueOfLife/backend/blob/master/dao/src/main/java/life/catalogue/es/nu/search/FiltersTranslator.java#L54

**Comment by ayco-at-naturalis** [comment]
@mdoering I think this is not going to work, but I'll look into it tomorrow. You now also wrap the filter on catalog key into the mustNot query, which is of course the opposite of what you want.

**Comment by mdoering** [comment]
hm, I thought this is exactly what I want. Tests look good. 
We need the catalogue key to be considered when looking for (not) existing decisions when the decision mode is given - no matter if its null, not null or some explicit value. Filtering on all decisions from all catalogues for a given usage does hardly ever make sense. The decision mode should always be accompanied by a catalogueKey which restricts the decision subdocument to consider in the query. I was even considering to validate requests to require a catalogue key when decision mode is given. But I thoght we dont really gain anything

**Comment by ayco-at-naturalis** [comment]
OK - I'll look again. Actually, I'm a bit baffled, because we're basically back to where I was (after I'd composensated for the decision being a nested object, and i **had** included the validation you mentioned above (in the RequestValidator)
